src/backend/dx11/src/lib.rs

```diff
@@ -1098,6 +1161,8 @@ impl hal::command::RawCommandBuffer<Backend> for CommandBuffer {
     }
 
     fn draw(&mut self, vertices: Range<VertexCount>, instances: Range<InstanceCount>) {
+        self.set_graphics_bind_point();
```

Comment:
that is something I strongly feel against - doing any kind of work on per-draw basis. This is the highest frequency we got, better try to rework this to happen at the PSO binding or something.

Reply:
Problem comes from DX11 pipeline states not holding vertex buffer strides and offsets and instead has to bind them on every `IASetVertexBuffer` call. This means we cant directly set the vertex buffers on a `set_vertex_buffer` call, but we have to wait until *after* our pipeline is bound and the only reliable place to set this is unfortunately at draw call. This can probably be improved quite a bit with some `dirty` bool flags, but I don't see a way around it.

Reply:
We actually have a similar problem in Metal (dependency of vertex buffers to the pipeline state). It is somewhat solved by trying to re-bind the vertex buffer on *BOTH* pipeline state setup and vertex buffer setup. There are cases where the combined state doesn't make sense (and we strongly expect the user to follow up with binding some other PSO/vertex buffer before issuing actual draw calls), and we should try to detect it here are avoid setting it up with D3D12 (bonus points for asserting if the user *does* try to render with the incomplete state).

Reply:
Would having a conditional jump to check for dirty state and set them appropriately be *that* bad to put in `draw_*`? There's a bit more state that needs to be separated as well (eg. scissors, viewports dont allow for providing an offset) and I can imagine it could get a bit messy. I'm thinking of something like the following:
```rust
fn set_state(&mut self) {
    if self.state.contains(DirtyState::VERTEX_BUFFERS) {
        self.bind_vertex_buffers();
    }

    ...

    self.state = 0;
}

fn draw(&mut self, ...) {
    if self.dirty != 0 {
        self.set_state()
    }

    unsafe {
        self.context.DrawInstanced(
            ...
        );
    }
}
```

Reply:
This dirty mechanics is what Angle and MoltenVK do. If you think of an app doing 10K-50K draw calls *per frame* (which is 1M+ per second), those little switches start matter. Not just in terms of the branch cost, but more about the code inlining and the instruction cache hits.
We could just give up on that sort of constraint in D3D11 and make it a truly 2nd tier backend, and ultimately it's your call to make, since you are the owner of D3D11 backend, but I feel responsible to provide you with the overall vision of an "ideal" backend that I myself at least try to reach in Metal/D3D12 backends.

Reply:
And yes, this may sound like a premature optimization. We may very well recognize later that the overhead of the driver is so high that our switches don't matter. It's crucial to measure this overhead, but we don't have any real-world benchmarks (other than Dota, perhaps). Changing this decision later along the way may also be expensive.

Reply:
You have me convinced :-)